### PR TITLE
Echo

### PR DESCRIPTION
**Pull request by jacquestardie**

### CODE REVIEW

File: echo_client.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import socket as s
+import sys
+
+def client(message):
+    message = message.encode('utf-8')
```
Comment (cewing) [inline echo_client.py]: Be a bit careful about just arbitrarily calling `encode` on the value passed as `message`.  What happens if it is already a byte string?  For example, consider this little interpreter session:

```python
>>> message = u"der übersetzer war tod"
>>> bytes = message.encode('utf-8')
>>> message
u'der \xfcbersetzer war tod'
>>> bytes
'der \xc3\xbcbersetzer war tod'
>>> message.encode('utf-8')
'der \xc3\xbcbersetzer war tod'
>>> bytes.encode('utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 4: ordinal not in range(128)
>>>
```
What is that UnicodeDecodeError coming from?  Could it happen to you on this line of code?

File: echo_server.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        message = ''
+        complete = False
+
+        while not complete:
+            piece = connection.recv(buffer)
+            if len(piece) < buffer:
+                complete = True
+            message = "{}{}".format(message, piece)
+
+        response = "{}{}".format("Received: ", message)
```
Comment (cewing) [inline echo_server.py]: This is not echoing the message the client sent.  It alters it and sends the altered version back.

File: echo_server.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        connection.sendall(response)
+        connection.close()
+    server.close()
```
Comment (cewing) [inline echo_server.py]: Be a bit careful about when you call this.  As it stands, if the server exits abnormally (like by doing a Keyboard Interrupt), this line will not be reached and your server socket may never be shut.

File: test_echo.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+def test_string():
+    '''Verify that messages are preserved in transmission.'''
+    test = client('Cats are kind of meh.')
+    assert test == 'Received: Cats are kind of meh.'
+
+def test_unicode():
+    '''Verify that unicode values are accepted.'''
+    test = client('♥'.encode('utf-8'))
+    assert test == u'Received: ♥'
```
Comment (cewing) [inline test_echo.py]: There are a few edge cases you haven't tested here.  I think at a minimum a good test suite for this setup requires testing a message shorter than one buffer size for the server and the client, a message longer than one buffer size for both, and a message that is exactly a multiple of the buffer size to verify that the process works even if the end of the message and the end of a single buffer received match up exactly.
